File: python/Binary Search/3.py
n,m = map(int, input().split())
arr = list(map(int, input().split()))

arr.sort()
sum = 0 # 얻은 떡의 합
max = 0 # 구하고자 하는 값 (높이의 최대값)

h = arr[n-1] # 현재 높이

# 현재 높이보다 큰 떡의 길이만 남기기 위한 이진탐색
def binary_search(arr, target, start, end):
    if start > end:
        return arr[start:]
    mid = (start + end) // 2
    if (arr[mid] == target):
        return arr[mid+1:]
    elif(arr[mid] > target):
        return binary_search(arr, target, start, mid - 1)
    else:
        return binary_search(arr, target, mid + 1, end)

while(max == 0):
    h -= 1 # 높이를 1씩 감소시키면서 확인
    sum = 0 
    temp = binary_search(arr, h, 0, n-1) # 높이보다 큰 떡들만 들어있는 배열 반환
    for i in temp:
        sum += i - h
        if sum >= m:
            max = h
            break

print(max)

# ---
# solve긴 하지만 시간초과 날 확률이 높은 코드.
# 떡의 길이들에서 이진탐색을 하는게 아니라 얻을 수 있는 떡의 합을 기준으로 이진탐색을 해야함.
# 왜? 높이의 범위가 10억까지니까 10억부터 1씩 감소시켜가는 코드는 너무 비효율적이다.
# 그래서 얻은 떡의 총합을 mid로 두고, 반씩 옮겨가면서 최적의 값을 찾음  (파라메트릭 서치를 이진탐색을 이용하여 해결)
# 입력값의 범위를 잘 보고, 이 코드가 시간제한을 맞출 수 있을지 생각해보기 필요

start = 0
end = max(arr)

while(start <= end):
    total = 0
    mid = (start + end) // 2
    for x in arr:
        #잘랐을때의 떡 양 계산
        if x > mid:
            total += x - mid
    # 떡의 양이 부족한 경우 더 많이 자름 (왼쪽 부분 탐색)
    if total < m:
        end = mid - 1
    # 떡의 양의 충분한 경우 덜 자르기 (오른쪽 부분 탐색)
    else:
        result = mid #최대한 덜 잘랐을경우를 구해야 하므로, 여기에서 result기록
        start = mid + 1

print(result)

# 4 6
# 19 15 10 17

File: 이취코테/javascript/Graph/3.js
import fs from "fs";
let input = fs.readFileSync("../test.txt").toString().split("\n");
const [n, m] = input[0].split(" ").map(Number);

const find = (parent, x) => {
  if (parent[x] !== x) parent[x] = find(parent, parent[x]);
  return parent[x];
};

const union = (parent, a, b) => {
  a = find(parent, a);
  b = find(parent, b);
  if (a < b) parent[b] = a;
  else parent[a] = b;
};

const parent = new Array(n + 1);
for (let i = 0; i < n + 1; i++) {
  parent[i] = i;
}

const edges = []; //간선 배열
let result = 0;
let maxResult = 0;

for (let i = 1; i < m + 1; i++) {
  const [a, b, c] = input[i].split(" ").map(Number);
  edges.push([c, a, b]);
}

edges.sort((x, y) => x[0] - y[0]); // 비용 기준으로 정렬

for (let edge of edges) {
  const [cost, a, b] = edge;
  //사이클이 발생하지 않는 경우에만 집합에 포함
  if (find(parent, a) !== find(parent, b)) {
    union(parent, a, b);
    result += cost;
    //✔ 크루스칼 알고리즘은 그리디 알고리즘으로, 간선이 최소인것부터 선택하기 때문에 max를 사용할 필요 없이 그냥 나중의 cost가 항상 최대가 된다.
    // maxResult = Math.max(maxResult, cost);
    maxResult = cost;
  }
}

console.log(result - maxResult); //비용 max인 간선 제거

//solve😀
//최소신장트리 알고리즘(크루스칼 알고리즘)을 그대로 사용하는 문제 (기억하고 있자) ✔
//최소신장트리 2개를 만드는 것이므로, 전체에서 최소신장트리를 만든 후에 가장 비용이 큰 간선을 제거하면 된다.

// 7 12
// 1 2 3
// 1 3 2
// 3 2 1
// 2 5 2
// 3 4 4
// 7 3 6
// 5 1 5
// 1 6 2
// 6 4 1
// 6 5 3
// 4 5 3
// 6 7 4
